### PR TITLE
Explicitly merge module augmentation members into the exports added by export * declarations

### LINKED ISSUE #35483: Expected Interface Augmentation, Overrode Interface Instead
Our behavior when augmenting a module that reexports other things is surprising. You end up accidentally overriding entire exports, rather than merging with them.

**Code**

```ts
// @filename: file.ts
export interface Foo {
    x: number;
}
// @filename: reexport.ts
export * from "./file";
// @filename: augment.ts
import * as ns from "./reexport";

declare module "./reexport" {
    export interface Foo {
        self: Foo;
    }
}

declare const f: ns.Foo;

f.x;
f.self;
f.self.x;
f.self.self;
```

**Expected behavior:**
No errors.

**Actual behavior:**
`Property 'x' does not exist on type 'Foo'.`

Workaround: Augment the original file containing the declaration in question, rather than the file that reexports it.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #35483

If this fix seems odd to anyone, we could also approach this by flagging the situation as likely an error (rather than merging into the targets of the export *), or by calling the current behavior "surprising but expected" and leaving it as-is. But given the mental model of augmentations (modifying the exports of another module), I think today's behavior of _not_ actually augmenting exports introduced by `export *`'s is a little _too_ surprising, so would rather see us change that behavior.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot user test this
@typescript-bot test this
@typescript-bot run dt

**Comment by andrewbranch** [comment]
I just tried this out locally, but I changed the real exported symbol to an enum, and the one in the module augmentation to a namespace, and it seems like the enum got overwritten. I then tried making both declarations enums, and then the module augmentation got shadowed. (I’m guessing the order was dependent on my exact editing scenario, and might be unstable.)

- Two interfaces worked fine, two namespaces did not.
- An enum and an interface raised the appropriate error.
- A class and an interface seemed to work

 Maybe there are issues when both symbols have a value meaning?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1251,6 +1251,15 @@ namespace ts {
...
+                        if (mainModule.exports?.get(InternalSymbolName.ExportStar) && moduleAugmentation.symbol.exports && moduleAugmentation.symbol.exports.size) {
+                            // We may need to merge the module augmentation's exports into the target symbols of the resolved exports
+                            const resolvedExports = getResolvedMembersOrExportsOfSymbol(mainModule, MembersOrExportsResolutionKind.resolvedExports);
+                            for (const [key, value] of arrayFrom(moduleAugmentation.symbol.exports.entries())) {
+                                if (resolvedExports.has(key) && !mainModule.exports.has(key)) {
+                                    mergeSymbol(resolvedExports.get(key)!, value);
+                                }
+                            }
+                        }
                         mergeSymbol(mainModule, moduleAugmentation.symbol);
```
Comment (sandersn) [inline src/compiler/checker.ts:1263]: remind me: what does it mean to still `mergeSymbol(mainModule, moduleAugmentation.symbol)` after manually merging the resolved export star? Are they mutually exclusive, so both need to be done?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1251,6 +1251,15 @@ namespace ts {
                         patternAmbientModuleAugmentations.set((moduleName as StringLiteral).text, merged);
                     }
                     else {
+                        if (mainModule.exports?.get(InternalSymbolName.ExportStar) && moduleAugmentation.symbol.exports && moduleAugmentation.symbol.exports.size) {
```
Comment (sandersn) [inline src/compiler/checker.ts:1254]:
```suggestion
                        if (mainModule.exports?.get(InternalSymbolName.ExportStar) && moduleAugmentation.symbol.exports?.size) {

```

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts:1263]: the manual merging is merging anything that exists in the export star and the augmentation into the export star target, the full ns merge here ensures that any augmentations that introduce new exports are then picked up and added.

**Comment by weswigham** [comment]
@typescript-bot user test this
@typescript-bot test this
@typescript-bot run dt

just in case~

**Pull request merged**